Clarify focus on promises in extensions APIs, cleanup (#1671)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/mv3/promises/index.md

---

> With the introduction of Manifest V3, many extension API methods now support promises. This document

Author: I can't think of a place where the product actually references manifests. Perhaps we do in some error messages, such as when failing to parse a manifest file. For the most part, though, this has been more of a marketing term than a product feature.

When I joined, I discussed a couple of different options with the team including "manifest version 3", "Maniest Version 3", "manifest v3", "Manifest v3", "Manifest V3", "mv3", and "MV3". We initially settled on "Manifest V3" as our primary name, but I also occasional used "MV3" in emails on the chromium-extensions group. When the topic of naming came up again in recent months, the extensions team's TLM and I discussed this and decided to settle on "Manifest V3" as our prefered (and only) term for all audiences.

> With the introduction of Manifest V3, many extension API methods now support promises. This document

Author: Happy to indulge. Adding this suggestion to the batch.

> With the introduction of Manifest V3, many extension API methods now support promises. This document

Reviewer: Cool. Can you please add this our [handbooks's word list](https://github.com/GoogleChrome/web.dev/blob/main/src/site/content/en/handbook/word-list/index.md).

> explains how to use promises when calling these methods.

Author: Adding this suggestion to a batch merge.

One nit that comes up as a result of this changes is that both sentences in this paragraph now end with the word "promises," which feels a bit repetitive. My writer's brain isn't really working at the moment, though, so I don't have a good idea on how to rephrase this passage to address the problem.

* Coding in a synchronous style for invoking asynchronous functions
* A simple "fork and join" syntax for invoking concurrent functions

In Manifest V2 extension developers could use libraries to "promisify" Chrome's extensions APIs.
With the introduction of Manifest V3 many of Chrome's extensions APIs can now return promises. The
Chrome team is also progressively adding promise support to additional APIs.

{% Aside 'gotchas' %}

> Promises are not available for extensions using Manifest V2, and are not available on all API

Reviewer 2: This still "[specifically says MV2 extensions cannot use Promises](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/3a3UzLbw1V4/m/FAmdT42UAAAJ)". Something like the following instead?

```suggestion
Promises are not returned by extension APIs under Manifest V2, ...
```

Reviewer: More like this:

"and are not yet available on all V3 methods."

Did you mean to put ellipses?

Reviewer 2: Just making a partial suggestion.

Author: Since there isn't a single suggestion that I can add to address this feedback, I'll be integrating these comments in a separate commit.

promise support on the method yet; in many cases it's because using a promise isn't feasible for the
method.

> You can check whether or not an API method supports promises by checking its API reference page:

Reviewer: ```suggestion
You can check whether a method supports promises by checking its API reference page:
```

{% Img src="image/SHhb2PDKzXTggPGAYpv8JgR81pX2/AYQVtnh19vNMHoXzxZB1.png",

>   alt="Screenshot showing a method that supports promises", width="800", height="280" %}

Author: Applying the suggested copy.

> The screenshot above shows a method from the [`chrome.tabs`][api-tabs-methods] API. You can see that

Reviewer 3: Rather than stating "The screenshot above", you can say:

> The example method, `captureVisibleTab`, can be found in the [`chrome.tabs`][api-tabs-methods] API.
> This method supports promises because one of the method's signatures returns a promise. ...

Author: Applying the suggested copy.

Reviewer 3: Awesome, thank you!

* Any time that you want to clean up your code by using a more "synchronous" invocation style.
* Where error handling would be too difficult using callbacks.

> * When you want a simpler way to invoke a number of concurrent methods and gather the results into a

Reviewer 3: [Avoid "simpler,"](https://developers.google.com/style/word-list#simple) as it may not be simple for all. Instead:

> * When you want to invoke a number of concurrent methods and gather the results into a single thread of code

Reviewer: It's not simple, it's simpler than the perviously described solution. It's not the same thing.

Reviewer 3: I had assumed it was related to it being "simpler" for the user, in which case I'd stand by removing it but maybe the suggestion was incorrect and should have suggested a synonym. Condensed? Shorter? 

If these are still off base, feel free to reject/ignore. It's something I noticed while reviewing open PRs, but it's your coverage/call.

Author: Revised to "When you want a more condensed way to invoke several concurrent methods and gather the results into a single thread of code."

  // When not wrapped in try/catch, errors thrown in an async
  // function will propagate down the promise chain
  let tabs = await chrome.tabs.query(queryOptions);
  if (!tabs.length) return;

  let tab = await chrome.tabs.create({
    url: 'https://example.com',
    index: tab[0].index + 1,
  });
  if (!tab) return;

  console.log('tab created', tab);
  return tab;
}
```

{% Aside %}

> Note that `await` is only valid in async functions and the top-level bodies of modules.

Reviewer 4: Could you add something like:

You can ensure your background Service Worker is run as a module by setting `"type": "module"` in the `"background"` key of your manifest.

With "manifest" linking to https://deploy-preview-1671--developer-chrome-com.netlify.app/docs/extensions/mv3/manifest/.
(That page hilariously has had `"type":` dropped from the `"module"` bit)?

Author: Applied your suggestion with some slight tweaks. Also fixed the manifest issues you mentioned.

  // When not wrapped in try/catch, errors thrown in an async
  // function will propagate down the promise chain
  let tabs = await chrome.tabs.query(queryOptions);
  if (!tabs.length) return;

  let tab = await chrome.tabs.create({
    url: 'https://example.com',
    index: tab[0].index + 1,
  });
  if (!tab) return;

  console.log('tab created', tab);
  return tab;
}
```

{% Aside %}

> Note that `await` is only valid in async functions and the top-level bodies of modules. You can load

Reviewer: ```suggestion
Note that `await()` is only valid in async functions and the top-level bodies of modules. You can load
```

Author: @jpmedley, I'm not sure this suggestion applies. In this case [`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await) is an operator that modifies an expression to await the resolution of a Promise. If this was intentional, would you mind explaining how it applies here?

Reviewer: Brain not working that day. Sorry.
